docs/source/building/variants.rst

```diff
@@ -0,0 +1,911 @@
+``exclude_from_build_hash`` entries from any of your conda_build_config.yaml
+files.
+
+
+Customizing compatibility
+-------------------------
+
+.. _pinning_expressions:
+
+Pinning expressions
+~~~~~~~~~~~~~~~~~~~
+
+
+Pinning expressions are the syntax used to specify how many parts of the version
+to pin. They are by convention strings containing ``x`` characters separated by
+``.``. The number of version parts to pin is simply the number of things that
+are separated by ``.``. For example, ``"x.x"`` pins major and minor version.
+``"x"`` pins only major version.
+
+Wherever pinning expressions are accepted, you can customize both lower and
+upper bounds.
+
+
+.. code-block:: python
+
+    # produces pins like >=1.11.2,<1.12
+    variants = [{'numpy': '1.11', 'pin_run_as_build': {'numpy': {'max_pin': 'x.x'}}}]
+
+
+.. code-block:: python
+
+    # produces pins like >=1.11,<2
+    variants = [{'numpy': '1.11', 'pin_run_as_build': {'numpy': {'min_pin': 'x.x', 'max_pin': 'x'}}}]
+
+
+Pinning at the variant level
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+Conda-build will automatically pin run requirements to the versions present in
+the build environment when the follow conditions are met:
+
+1. The dependency is listed in the requirements/build section. It can be pinned,
+   but does not need to be.
+2. The dependency is listed by name (no pinning) in the requirements/run section
+3. The ``pin_run_as_build`` key in the variant has a value that is a dictionary,
+   containing a key that matches the dependency name listed in the run
+   requirements. The value should be a dictionary with up to 4 keys:
+   ``min_pin``, ``max_pin``, ``lower_bound``, ``upper_bound``. The first two are
+   pinning expressions. The latter two are version numbers, overriding detection
+   of current version. (defaulting to None).
+
+An example variant/recipe is shown here:
+
+conda_build_config.yaml:
+
+.. code-block:: yaml
+
+    boost: 1.63
+    pin_run_as_build:
+        boost:
+          max_pin: x.x
+
+meta.yaml:
+
+.. code-block:: yaml
+
+   requirements:
+       build:
+           - boost {{ boost }}
+       run:
+           - boost
+
+
+The result here is that the runtime numpy dependency will be pinned to
+``>=(current boost 1.63.x version),<1.64``
+
+Note that there are some packages that you should not use ``pin_run_as_build``
+for. Packages that don't *always* need to be pinned should be pinned on a
+per-recipe basis (described in the next section).  Numpy is an interesting
+example here. It actually would not make a good case for pinning at the variant
+level. Because you only need this kind of pinning for recipes that use Numpy's C
+API, it would actually be better to not pin numpy with ``pin_run_as_build``.
+Pinning it is over-constraining your requirements unnecessarily when you are not
+using Numpy's C API. Instead, we should customize it for each recipe that uses
+numpy.  See also the `Avoiding unnecessary builds`_ section above.
```

Comment:
It would be helpful to have a clear statement at the beginning of this section that `pin_run_as_build` is intended only for those cases when all packages should be pinned (it seems like the most familiar case would be python if you are building python packages).